common/io.hpp

```diff
@@ -89,9 +90,9 @@ class StreamReader : virtual public Reader {
 	ExpectedSize Read(vector<uint8_t>::iterator start, vector<uint8_t>::iterator end) override {
 		is_.read(reinterpret_cast<char *>(&*start), end - start);
 		if (is_.bad()) {
-			int int_error = errno;
-			return Error(
-				std::error_code(int_error, std::system_category()).default_error_condition(), "");
+			int io_error = errno;
+			return expected::unexpected(
+				Error(std::generic_category().default_error_condition(io_error), strerror(errno)));
```

Comment:
No need for `strerror` here, you are just duplicating the message.

Reply:
We don't have a good way to extract the message from the `error_condition`, AFAICT. And when processing our errors as part of the `ExpectedSomething` types, we do `ex.error().message`. So either the `Error` constructor should do the something like `message(strerror(the_extracted_errno))` or we should do it explicitly like above.

Reply:
Ah, but we do! Usually we don't look at `message` directly, but use [`String()`](https://github.com/mendersoftware/mender/blob/feature-c%2B%2B-client/common/error.hpp#L60), which combines both messages.

Reply:
> Ah, but we do! Usually we don't look at `message` directly, but use [`String()`](https://github.com/mendersoftware/mender/blob/feature-c%2B%2B-client/common/error.hpp#L60), which combines both messages.

Oh, right! I'm not sure we do use `String()` consistently, but we definitely should. I'll update this.